zephyr: remove deprecated Kconfigs

ycsin opened the pull request:
These Kconfigs should have been deprecated for more than 2 releases, remove them:

- `CONFIG_COUNTER_RTC_STM32_CLOCK_LSI`
- `CONFIG_COUNTER_RTC_STM32_CLOCK_LSE`
- `CONFIG_EMUL_EEPROM_AT2X`
- `CONFIG_ETH_STM32_HAL_RANDOM_MAC`
- `CONFIG_ETH_STM32_HAL_USER_STATIC_MAC`
- `CONFIG_ETH_STM32_HAL_MAC3`
- `CONFIG_ETH_STM32_HAL_MAC4`
- `CONFIG_ETH_STM32_HAL_MAC5`
- `CONFIG_IEEE802154_2015`
- `CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ_RADIO_TX_RETRIES`
- `CONFIG_CPP_MAIN`
- `CONFIG_CPLUSPLUS`
- `CONFIG_LIB_CPLUSPLUS`
- `CONFIG_EXCEPTIONS`
- `CONFIG_RTTI`
- `CONFIG_SUPPORT_MINIMAL_LIBC`
- `CONFIG_MINIMAL_LIBC_MALLOC`
- `CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE`
- `CONFIG_MINIMAL_LIBC_CALLOC`
- `CONFIG_MINIMAL_LIBC_REALLOCARRAY`
- `CONFIG_PICOLIBC_HEAP_SIZE`
- `CONFIG_BT_DEBUG_LOG`
- `CONFIG_LOG_DOMAIN_ID`
- `CONFIG_NET_L2_IEEE802154_ACK_REPLY`
- `CONFIG_SETTINGS_FS`
- `CONFIG_SETTINGS_FS_DIR`
- `CONFIG_SETTINGS_FS_FILE`
- `CONFIG_SETTINGS_FS_MAX_LINES`

Remove related code in .c/.h as well.

keith-packard commented on lib/libc/Kconfig:
```diff
@@ -79,6 +72,9 @@ config MINIMAL_LIBC
 	imply COMPILER_FREESTANDING
 	select COMMON_LIBC_ABORT
 	select COMMON_LIBC_STRNLEN
+	imply COMMON_LIBC_MALLOC
```
Until #65434 is merged, I don't think we should be adding malloc to the default minimal libc configuration as that could increase the size of applications which don't use it. That's because the malloc heap initializer will always end up linked into applications because the common C library bits are included under the `--whole-archive` list.

cfriedt replied:
I had a change to remove the sys_init from the minimal libc malloc (common, I guess) and use lazy init instead (init on first use), but it was more complicated than that.

Honestly, that should have a bug / issue number attached to it, because it needs to be fixed. In no universe should we be unconditionally linking malloc buffers and code without an explicit call to malloc or something related.

ycsin replied:
This is moved from:

```
...

if MINIMAL_LIBC

...

config MINIMAL_LIBC_MALLOC
	bool "[DEPRECATED] Minimal libc malloc implementation"
	default y
	imply COMMON_LIBC_MALLOC
	help
	  [DEPRECATED] Use COMMON_LIBC_MALLOC

	  Enable the minimal libc's implementation of malloc, free, and realloc.
	  Disable if you wish to provide your own implementations of these functions.

...

```

which I have deleted in `lib/libc/minimal/Kconfig`

keith-packard replied:
Yup, but this changes the default. Before, you didn't get malloc unless you asked for it. With the proposed change, you get malloc unless you ask to *not* have it. Unless I'm mis-reading the patch?

ycsin replied:
Currently, enabling `MINIMAL_LIBC` would enable `MINIMAL_LIBC_MALLOC` by default, which would `imply` `COMMON_LIBC_MALLOC`, so malloc will be enabled unless user explicitly disables `MINIMAL_LIBC_MALLOC` and/or `COMMON_LIBC_MALLOC`.

FWIU, this PR removes the intermediary `MINIMAL_LIBC_MALLOC` & relocate the `imply COMMON_LIBC_MALLOC` directly under the `MINIMAL_LIBC`, so `COMMON_LIBC_MALLOC` will be enabled by default, which is the same as before, unless explicitly disabled by the user, but there's one less Kconfig (`MINIMAL_LIBC_MALLOC`) that a user has to care.

Or maybe I missed something?

keith-packard replied:
Nope, *I* missed that minimal libc sets `MINIMAL_LIBC_MALLOC` to `default y`. Thanks for your patience, this looks fine.

keith-packard commented on lib/libc/common/Kconfig:
```diff
@@ -53,7 +53,6 @@ config COMMON_LIBC_MALLOC_ARENA_SIZE
 config COMMON_LIBC_CALLOC
 	bool "Common C library calloc"
 	depends on COMMON_LIBC_MALLOC
-	default n if MINIMAL_LIBC && !MINIMAL_LIBC_CALLOC
```
You can just remove `&& !MINIMAL_LIBC_CALLOC` here. The goal is to disable calloc when using the minimal C library by default.

ycsin replied:
updated

keith-packard commented on lib/libc/common/Kconfig:
```diff
@@ -62,7 +61,6 @@ config COMMON_LIBC_CALLOC
 config COMMON_LIBC_REALLOCARRAY
 	bool "Common C library reallocarray"
 	depends on COMMON_LIBC_MALLOC
-	default n if MINIMAL_LIBC && !MINIMAL_LIBC_REALLOCARRAY
```
Same here -- leave the line, but remove `&& !MINIMAL_LIBC_REALLOC_ARRAY`

ycsin replied:
updated

keith-packard reviewed (changes requested):
Need to fix the semantic changes to the common C library malloc code when using the minimal C library.

erwango commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -627,11 +627,7 @@ static const struct stm32_pclken rtc_clk[] = {
 	STM32_CLOCK_INFO(0, DT_DRV_INST(0)),
 	/* Use Kconfig to configure source clocks fields (Deprecated) */
 	/* Fortunately, values are consistent across enabled series */
-#ifdef CONFIG_COUNTER_RTC_STM32_CLOCK_LSE
-	{.bus = STM32_SRC_LSE, .enr = RTC_SEL(1)}
-#else
 	{.bus = STM32_SRC_LSI, .enr = RTC_SEL(2)}
```
Please remove the whole `#if DT_INST_NUM_CLOCKS(0) == 1` section.

ycsin replied:
removed

erwango reviewed (changes requested):
Otherwise LGTM, thanks

erwango commented on drivers/counter/Kconfig.stm32_rtc:
```diff
@@ -22,20 +22,6 @@ choice COUNTER_RTC_STM32_CLOCK_SRC
-	  Use LSE as RTC clock
-	  Deprecated in favor of device tree secondary domain clock
-
 endchoice #COUNTER_RTC_STM32_CLOCK_SRC
```
Please remove the while choice

ycsin replied:
removed

erwango reviewed (changes requested):
Sorry for not seeing this earlier.

aescolar commented:
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

nashif commented:
You probably want to do this in multiple commits, remove unrelated Kconfigs all in one commit does work well and will make reviews difficult.

nashif commented:
this could go into 3.7 IMO

ycsin committed:
drivers: counter: stm32: remove deprecated clock source Kconfigs

These clock selection Kconfigs should have been deprecated for
more than 2 releases, remove them:

- `CONFIG_COUNTER_RTC_STM32_CLOCK_SRC`
- `CONFIG_COUNTER_RTC_STM32_CLOCK_LSI`
- `CONFIG_COUNTER_RTC_STM32_CLOCK_LSE`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
bluetooth: remove CONFIG_BT_DEBUG_LOG

`CONFIG_BT_DEBUG_LOG` has been deprecated for more than 2
releases, replace it with `CONFIG_BT` + `CONFIG_LOG`:

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
log: remove CONFIG_LOG_DOMAIN_ID

`CONFIG_LOG_DOMAIN_ID` has been deprecated for years and
shouldn't be used according to its documentation, remove it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: eeprom: remove CONFIG_EMUL_EEPROM_AT2X

`CONFIG_EMUL_EEPROM_AT2X` has been deprecated for more than 2
releases, remove it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: ethernet: stm32: remove deprecated MAC address Kconfigs

The following Kconfigs:

- `CONFIG_ETH_STM32_HAL_RANDOM_MAC`
- `CONFIG_ETH_STM32_HAL_USER_STATIC_MAC`
- `CONFIG_ETH_STM32_HAL_MAC3`
- `CONFIG_ETH_STM32_HAL_MAC4`
- `CONFIG_ETH_STM32_HAL_MAC5`

have been deprecated for more than 2 releases, user should
have switched to the devicetree method now, remove them.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: interrupt controller: update GIC Kconfigs

Selection of these GIC Kconfigs have been deprecated
for more than 2 releases, users should use the devicetree
method instead, update the Kconfigs.

The SOCs below have been updated to not select `GIC_V3`, since
their devicetree already have the required compatible:
- fvp_aemv8r
- rzt2m
- rk3658

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
lib: hash: update the include headers

Include "stdio.h" as `malloc()` is used in the header.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
net: ieee802154: remove CONFIG_NET_L2_IEEE802154_ACK_REPLY

`CONFIG_NET_L2_IEEE802154_ACK_REPLY` has been deprecated for
more than 2 releases, remove it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
settings: remove CONFIG_CONFIG_SETTINGS_FS

`CONFIG_CONFIG_SETTINGS_FS` has been deprecated for
more than 2 releases, remove it and its associated source
code.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
lib: cpp: remove deprecated Kconfigs

These Kconfigs should have been deprecated for more than 2
releases, remove them:

- `CONFIG_CPP_MAIN`
- `CONFIG_CPLUSPLUS`
- `CONFIG_LIB_CPLUSPLUS`
- `CONFIG_EXCEPTIONS`
- `CONFIG_RTTI`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

omkar3141 reviewed (dismissed):
Mesh changes look non-controversial.

ycsin committed:
driver: ieee802154: remove deprecated Kconfigs

The following Kconfigs have been deprecated for more than 2
releases, remove them:

- `IEEE802154_CC13XX_CC26XX_SUB_GHZ_RADIO_TX_RETRIES`
- `CONFIG_IEEE802154_2015`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
libc: minimal: remove deprecated Kconfigs

The following Kconfigs have been deprecated for more than 2
releases, remove them:

- `CONFIG_SUPPORT_MINIMAL_LIBC`
- `CONFIG_MINIMAL_LIBC_MALLOC`
- `CONFIG_MINIMAL_LIBC_CALLOC`
- `CONFIG_MINIMAL_LIBC_REALLOCARRAY`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
libc: cleanup deprecated heap-related Kconfigs

The following heap-related Kconfigs shared by the picolibc &
minimal libc have been deprecated for more than 2 releases,
remove them and update the Kconfigs accordingly:

- `CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE`
- `CONFIG_MINIMAL_LIBC_REALLOCARRAY`

Cleanup the handling to support the deprecated malloc arena
size configuration values.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
Split the commit into multiple commits, PTAL, thanks.

keith-packard reviewed (approved):
I missed that minimal C library defaults to malloc enabled. My previous comments were mis-informed.

stephanosio reviewed (approved):
libc and cpp changes LGTM.
